### ISSUE TITLE
Passing tensor to c++ extension function changes its options

### ISSUE THREAD

**Issue by serhii-havrylov** [pytorch/pytorch#17849]
## 🐛 Bug ?

Hello,

I have very simple c++ function:
```cpp
#include <torch/extension.h>
#include <iostream>

void print_options(torch::Tensor x) {
    std::cout << x.options() << "\n";
}

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
  m.def("print_options", &print_options, "");
}

```
which I use via JIT compiling extensions functionality:
```python
import torch
from torch.utils import cpp_extension

cuda_module = cpp_extension.load(name="cpp_module", sources=["cpp_modue/my_fun.cpp"])
print_options = cuda_module.print_options

a = torch.zeros(2, 3)
a.requires_grad = True
print(a.requires_grad)

print_options(a)
```
And here is the output:

```
True
TensorOptions(dtype=float, device=cpu, layout=Strided, requires_grad=false)
```

Is this expected behaviour? Why does passing tensor to the c++ function change its options (requires_grad in particular)?

 - PyTorch Version is 1.0.1

**Comment by t-vi** [comment]
Yes, this is expected behaviour. It's not the passing to C++ that does it, but rather the definition of `Tensor::options()`, which only reflects selected parts of the options. You need to consult `Tensor::requires_grad()` if you want to determine whether it requires grad or not.
This is on purpose: The intended use for `Tensor::options()` is to create new tensors compatible with existing ones (same device, dtype, ...), but inheriting `requires_grad` does not, in general make sense.

**Comment by serhii-havrylov** [comment]
Hi @t-vi! 
Thank you for clarification. Don't you think that this behaviour can be somewhat misleading sometimes? For example:

```c++
#include <torch/extension.h>
#include <iostream>


void print_options(torch::Tensor x) {
    auto dim1 = x.size(0);
    auto dim2 = x.size(1);
    auto a = torch::empty({dim1, dim2}, x.options().requires_grad(true));
    std::cout << a.options() << "\n";
    std::cout << a.requires_grad() << "\n";
}

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
  m.def("print_options", &print_options, "");
}
```
this code will produce the output:

```
TensorOptions(dtype=float, device=cpu, layout=Strided, requires_grad=false)
true
```

It looks like `requires_grad` changed its value between two calls (.options() and requires_grad()).
Is there a method that returns "real" option values?

**Comment by ezyang** [comment]
The behavior here looks wrong; if we have a TensorOptions that requires grad, we should get a requires tensor out.. So that looks like a legit bug.

I'm also pretty unhappy with `x.options()`. It's very unclear what intended semantics (does it include requires grad or not) are.

**Comment by t-vi** [comment]
First: sorry, for prematurely closing this!

Of course, Edward is much more knowledgeable than me here. That said, I'm not sure what you see as the use case.

When I do `b = zeros_like(a)`, I get a non-grad-requiring tensor, and I think that `torch::zeros(a.sizes(), a.options())` is and should be the C++-equivalent.
The semantics as I always understood them were just that: *`t.options()` gives you options that make the tensor useful for use with t*.
I can see how the print example is unfortunate, but the behaviour itself would seem like the most useful, because `requires_grad` isn't something you need to inherit to a new tensor (unlike device, ...).

**Comment by serhii-havrylov** [comment]
I guess I just was lucky enough to work on the case that belongs to that 1% and I wrongly assumed that `options()` would return precisely the option values that tensor contains without any modifications.

**Comment by t-vi** [comment]
Out of curiosity, what is the use case?

**Comment by serhii-havrylov** [comment]
I use the c++ extensions functionality to implement some custom autograd.Function. In the code of the forward pass I allocate a tensor which later is going to be filled in by some cuda kernel. It would be convenient if the `requires_grad` variable depends on the `requires_grad` value of the input tensor. Now, I can do that with 
```c++
auto a = torch::empty({dim1, dim2}, x.options().requires_grad(x.requires_grad()));
```
but it looks somewhat ugly, tbh.

**Comment by t-vi** [comment]
I'm slow to understand things today, but if you are wrapping it in an `autograd.Function`, would you have to set `requires_grad` yourself? A (nonsense) example:
```python
import torch
class X(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a):
        r = torch.zeros_like(a)
        print ("rg inner", r.requires_grad)
        return r
    @staticmethod
    def backward(ctx, b):
        return b

a = torch.randn(5, requires_grad=True)
b = X.apply(a)
print ("rg out", b.requires_grad)
```
will give
```
rg inner False
rg out True
```
So the requires_grad is set by the autograd function, not the creation of the tensor, which - maybe just because that's how I've worked with it - is what I would expect (it also means that you cannot easily output two floating tensors with one requiring grad and one not, but hey).

**Comment by serhii-havrylov** [comment]
You are right, `autograd.Function` looks after all those `require_grad`, my bad. For some reason, I assumed that I have to do that manually.

> it also means that you cannot easily output two floating tensors with one requiring grad and one not, but hey

I believe you can, e.g.:

```python
class MyFun(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = 2 * x
        z = 3 * x
        ctx.mark_non_differentiable(z)
        return y, z

    @staticmethod
    def backward(ctx, dy, dz):
        print(dy)
        print(dz)
        return 2 * dy, None


x = torch.tensor([2., 3.], requires_grad=True)
y, z = MyFun.apply(x)
print(y.requires_grad)
print(z.requires_grad)
loss = z.sum() + y.sum()
loss.backward()
print(x.grad)
print(y.grad)
```
To my surprise, I found out that `dz` is a zero tensor. I expected None, hmm 🤔.

**Comment by t-vi** [comment]
Ah right. :)

**Comment by SsnL** [comment]
@serhii-havrylov Yes that is a bug. See https://github.com/pytorch/pytorch/issues/12631

**Comment by ezyang** [comment]
By the way, the problem with TensorOptions is that the class was originally designed as a C++ idiomatic class to pass keyword arguments like `device` and `requires_grad` to our factory functions, but someone (quite reasonably) added `x.options()` and then we couldn't agree if this meant "exactly the set of options that would have been used to construct this tensor" or "the device, layout and dtype of the tensor, which is probably more likely to be what you want because you rarely want to clone the requires grad status of a tensor if you use options."

**Comment by gchanan** [comment]
I think putting `requires_grad` on TensorOptions was a mistake -- as @ezyang points out, the meaning is ambiguous, and it's trivial to just set it after the fact if you want it with `torch::factory_fn(...).requires_grad_()`.  This wasn't possible at some point in the past, because we had too strict checking (that the version counter of a SavedVariable was 0 in some places), and the above gives you are version counter of 1, but that has been fixed.

I'm not quite sure how to deal with backwards compatibility concerns, though.

**Comment by ezyang** [comment]
cc @VitalyFedyunin
